### PR TITLE
[Trigger Service/Oauth2-Middleware] Hocon config refactor

### PR DESCRIPTION
**Pull request by akshayshirahatti-da**
This adds a module `pureconfig-utils` which can have shared config readers and config case classes required for `oauth2-middleware`,`trigger-service` and `json-api`.

Also cleans up some of the oauth2-middleware code based on learnings from the trigger service code review.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by S11001001** [comment]
`main` base => @akshayshirahatti-da doesn't want this reviewed until its basis is merged

`hocon_trigger_service` => @akshayshirahatti-da won't merge the base's PR until this one is reviewed and merged

or this is my assumption from these choices, anyway.

**Comment by akshayshirahatti-da** [comment]
yeah the general idea was to have `hocon_trigger_service` merged before we review this, anyways I have fixed this to include only the single commit that matters please have a look @S11001001 @stefanobaghino-da

File: triggers/service/auth/src/main/scala/com/daml/auth/middleware/oauth2/Cli.scala
Comment (stefanobaghino-da) [inline triggers/service/auth/src/main/scala/com/daml/auth/middleware/oauth2/Cli.scala:27]: It doesn't look like this is a `case class`. There plenty of behavior encapsulated in this, including logging.

**Review by stefanobaghino-da — COMMENTED** [review]
No blockers from me, thanks.

File: triggers/service/auth/src/main/scala/com/daml/auth/middleware/oauth2/Cli.scala
Comment (S11001001) [reply triggers/service/auth/src/main/scala/com/daml/auth/middleware/oauth2/Cli.scala:27]: I think the important question is "identity", not "behavior". In other words, if `expr: C`, and for all `f`, `f(expr) = f(expr.copy())`, then `C` is a `case class`. Less formally, a case class is exactly the sum of its parts, which seems to hold here.

All of these methods can be written as auxiliary functions; their presence in the class body shouldn't change how we interpret the class's semantics.

File: triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/CliSpec.scala
Comment (S11001001) [inline triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/CliSpec.scala:72]:
```suggestion
      c.tokenVerifier shouldBe a[JwksVerifier]
```

**Review by S11001001 — APPROVED** [review]

**Pull request merged**
